lib/redis.rb

```diff
@@ -2130,13 +2130,14 @@ def unwatch
 
   def pipelined
     synchronize do |client|
+      pipeline = Pipeline.new
       begin
-        original, @client = @client, Pipeline.new
+        original, @client = @client, pipeline
         yield(self)
-        original.call_pipeline(@client)
       ensure
         @client = original
       end
+      original.call_pipeline(pipeline)
```

Comment:
Ah, yes, I understand it now.
I would have liked if instead of modifying the client directly, we would have an API like:

```ruby
redis.pipelined do |pipe|
  pipe.hgetall(:key)
end
```
without affecting the outer Redis object, sadly because of backwards-compatibility we need to keep the current one (maybe I can think of a good way to extend it)